src/Data/Generic.purs

```diff
@@ -49,6 +51,50 @@ data GenericSignature = SigProd String (Array DataConstructor)
                       | SigChar
                       | SigArray (Unit -> GenericSignature)
 
+showSignature :: GenericSignature -> String
+showSignature sig =
+  fold $ case sig of
+    SigProd tyName ctors ->
+      ["SigProd ", show tyName, " ", showArray showDataConstructor ctors]
+    SigRecord labels ->
+      ["SigRecord ", showArray showLabel labels]
+    SigNumber ->
+      ["SigNumber"]
+    SigBoolean ->
+      ["SigBoolean"]
+    SigInt ->
+      ["SigInt"]
+    SigString ->
+      ["SigString"]
+    SigChar ->
+      ["SigChar"]
+    SigArray sig ->
+      ["SigArray (", showSignature (force sig), ")"]
```

Comment:
This could result in additional parens, but no big deal. We can resolve it by splitting out things which don't need to be parenthesized into their own `case`, if need be.